---
title: "methylscaper: Interactive Visualization of Methylation and Nucleosome Occupancy Data"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteEngine{knitr::knitr}
  %\VignetteIndexEntry{methylscaper: Interactive Visualization of Methylation and Nucleosome Occupancy Data}
  %\usepackage[UTF-8]{inputenc}
---

# Introduction

`methylscaper` is an R package that provides functions for processing and
visualizing data generated by methods jointly profiling methylation and chromatin accessibility (MAPit, NOMe-seq, scNMT-seq, nanoNOMe, etc.). The package offers processing for
both single-cell and single-molecule data, and a common interface for jointly visualizing both data types through the generation of ordered representational methylation-state matrices. Users may also run the package through a Shiny app which provides an interactive seriation
process of refinement and re-weighting that optimally orders the single DNA molecules to discover methylation patterns and nucleosome positioning.

<!-- If you use methylscaper in your research please cite our preprint on bioRxiv: [methylscaper manuscript](https://www.biorxiv.org/content/10.1101/2020.11.13.382465v1) -->

<!-- If after reading this vignette, you have questions, please submit your question on GitHub: [Question or Report Issue](https://github.com/rhondabacher/methylscaper/issues). This will notify the package maintainers and benefit other users. -->


<!-- ## Installation -->

<!-- `methylscaper` can be installed via GitHub. -->

<!-- ```{r, eval = FALSE} -->
<!-- install.packages(devtools) -->
<!-- devtools::install_github("rhondabacher/methylscaper", build_vignettes=TRUE) -->
<!-- library(methylscaper) -->
<!-- ``` -->

<!-- After successful installation, you can load the package into the working space: -->
<!-- ```{r, echo=FALSE, message = FALSE} -->
<!-- library(methylscaper) -->
<!-- ``` -->

<!-- To access the shiny app, simply run: -->
<!-- ```{r, eval=F} -->
<!-- methylscaper() -->
<!-- ``` -->

<!-- # Usage -->

<!-- ## Single-cell data -->

<!-- For analyzing single-cell data from methods such as scNMT-seq, methylscaper begins with pre-aligned data. The data should be in the form of three columns (chromosome, position, methylation status). There should be two files for each cell, one for the GCH sites and another for the HCG sites. This file can be generated automatically via the "bismarck\_methylation\_extractor" script in the Bismarck software tool.  -->

<!-- Methylscaper then has a function to process this data before the visualization analysis. The files should be organized such that all GCH files are location in a folder labelled GC, and all HCG files in a folder labelled CG. In the shiny app, first point to the directory containing these two subdirectories. Then for optimal processing speed, we usually filter to the chromosome level, however, if users wish to further subset the data then a start and end can be specified here as well. We advise users to keep the region relatively large as the next step allows you to search along the chromosome. -->

<!-- ### Example workflow for single-cell data -->

<!-- Below we walk through an example using data from Clark et al., 2018, obtained from -->
<!-- [GSE109262](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE109262). For -->
<!-- the sake of this example, we assume that the `GSE109262_RAW.tar` -->
<!-- directory is downloaded to `~/Downloads/`. After extraction we then created the GC and CG subdirectories. -->

<!-- In the screenshot below, we select chromosome 19, and for the purpose of creating a small example we restricted between sites XX and XX centered around the Eef1g gene. After this processing is done, the shiny app will automatically move the files to the visualization tab for analysis. -->

<!-- ##screenshot here -->

<!-- The preprocessing and visualization can also be done outside the shiny app in the R console directly.  -->

<!-- Pre-processing: -->
<!-- ```{r, eval=FALSE} -->
<!-- ## Some code here  -->
<!-- prepsc.out <- prepSC(gc.seq.sub, cg.seq.sub, startPos = 8966841, endPos = 8967541) -->
<!-- gch <- prepsc.out$gch -->
<!-- hcg <- prepsc.out$hcg -->
<!-- ``` -->

<!-- The `gch` and `hcg` objects are matrices representing accessibility -->
<!-- and methylation status, respectively, and are used by other -->
<!-- `methylscaper` functions for visualization. -->

<!-- Visualization: -->

<!-- The `initialOrder` function computes an ordering of the state matrices, using a given method. By default, the function uses a PCA-based ordering which we find optimal and efficiencly scales to large datasets, but technically any method supported by the `seriation` package could be used. It can also perform a weighted ordering, according to the methylation or accessibility status of a subset of the data's columns. -->


<!-- ```{r, eval=FALSE} -->
<!-- orderObj <- initialOrder(input.GCH = gch, input.HCG = hcg,  -->
<!--                          weightStart = 47, weightEnd = 358, weightFeature = "yellow") -->
<!-- ``` -->

<!-- `initialOrder` returns an object of class `orderObject`, which -->
<!-- includes the combined state data matrix and the ordering of the -->
<!-- reads. We then generate a sequence plot with the `plotSequence` -->
<!-- function. The option 'plotFAST' generated a .png image. To save a high resolution PDF, change 'plotFAST' to TRUE. -->

<!-- The plot on the left corresponds to methylation status, and the right to -->
<!-- accessibility status The red and yellow colored portions represent the areas on each -->
<!-- read that are methylated and accessible, respectively. -->

<!-- ```{r, eval=FALSE} -->
<!-- plotSequence(orderObj, Title = "Eef1g gene", plotFAST=FALSE) -->
<!-- ``` -->


<!-- We can also refine the ordering of the reads with `refineFunction`, -->
<!-- which reorders a subset of the reads with a given method. The -->
<!-- code below reorders the first 50 reads and generates a new sequence plot. -->

<!-- ```{r, eval=FALSE} -->
<!-- orderObj$order1 <- refineFunction(orderObj, refineStart = 1, refineEnd = 50) -->
<!-- plotSequence(orderObj, Title = "Eef1g gene", plotFAST=FALSE) -->
<!-- ``` -->





<!-- ### Single-molecule data -->

<!-- For single-molecule data from MAPit experiments, methylscaper is able to align reads from a fasta file to an appropriate reference file. This can also be done in both the Shiny app and in the R console. -->

<!-- For the Shiny app, the input should be a list of reads in a fasta format and a fasta reference file. After alignment, the app with shift to the analysis tab where the output can be downloaded directly. The screenshot below shows both steps. -->


<!-- To run the preprocessing in teh R console, the function `runAlign` may be used. The sequences are aligned to the reference -->
<!-- with the `Biostrings` package, and then mapped to the methylation- and -->
<!-- accessibility-state matrices. For very large datasets, the align function may be run on high-throughput servers rather than locally. -->

<!-- ```{r, eval = FALSE} -->
<!-- fasta <- read.fasta("seq_file.fasta") -->
<!-- ref <- read.fasta("ref.Pacbio.fa") -->
<!-- align.out <- runAlign(fasta = fasta, ref = ref) -->
<!-- gch <- align.out$gch -->
<!-- hcg <- align.out$hcg -->
<!-- ``` -->

<!-- The visualization analysis is then done by: -->

<!-- ```{r, eval=FALSE} -->
<!-- bc4.order.pca <- initialOrder(gch, hcg,  -->
<!--                               Method="PCA", weightStart = 308, weightEnd = 475, weightFeature = "red") -->
<!-- plotSequence(bc4.order.pca, Title = "Ordered by PCA", plotFAST = T) -->
<!-- bc4.order.pca$order1 <- refineFunction(bc4.order.pca, refineStart = 54, refineEnd = 1, Method = "PCA") -->
<!-- plotSequence(bc4.order.pca, Title = "Ordered by PCA", plotFAST = T) -->
<!-- ``` -->
<!-- The analysis plots are also available directly: -->

<!-- ```{r, eval=FALSE} -->
<!-- percent_C(orderObject = bc4.order.pca, plotPercents = TRUE, main="Percent of methylated reads") -->

<!-- proportion_color(bc4.order.pca, plotHistogram = TRUE, color = "YELLOW", main="GC methylation proportion") -->

<!-- proportion_color(bc4.order.pca, plotHistogram = TRUE, color = "RED", main="CG methylation proportion") -->
<!-- ``` -->


<!-- ## Example analysis -->

<!-- To demonstrate the visualization, we have included a pre-processed object available called `day7`. -->

<!-- ```{r} -->
<!-- data(day7) -->
<!-- gch <- day7$gch -->
<!-- hcg <- day7$hcg -->
<!-- orderObj <- initialOrder(gch, hcg, Method = "PCA") -->
<!-- ``` -->


<!-- ```{r} -->
<!-- plotSequence(orderObj) -->
<!-- ``` -->


<!-- ```{r} -->
<!-- orderObj$order1 <- refineFunction(orderObj, 1, 50) -->
<!-- plotSequence(orderObj) -->
<!-- ``` -->

<!-- We can also generate basic summary plots of the data. -->

<!-- ```{r, fig.height = 5, fig.width = 6, fig.align="center"} -->
<!-- percent.c.out <- percent_C(orderObj, plotPercents = TRUE) -->
<!-- prop.color.out <- proportion_color(orderObj, color = "YELLOW", plotHistogram = TRUE) -->
<!-- ``` -->



<!-- ## Shiny App -->

<!-- The Shiny app offers all of the package functionality described above, and can be run -->
<!-- by calling the function `methylscaper()` from an R session with the -->
<!-- package loaded. The app is organized into single-cell and -->
<!-- single-molecule panels, which each have tabs for processing and -->
<!-- visualization. -->

<!-- In the "Single-cell" panel, users upload the raw data as described above. After the initial -->
<!-- processing is complete, the Shiny app generates a slider that can be -->
<!-- used to adjust the selected region.  -->

<!-- Once the initial sequence plot is generated, `methylscaper` allows the -->
<!-- user to dynamically refine and re-weight the plot via Shiny's brushing -->
<!-- mechanics. Clicking and dragging along either of the two plots will -->
<!-- select sites (i.e., columns) by which to weight the data in the site -->
<!-- matrix. `methylscaper` will then regenerate the plot with a new -->
<!-- ordering, influenced by the weighted sites. With "PCA" selected as the -->
<!-- seriation method, the new ordering will be generated with a weighted -->
<!-- Principal Components Analysis. If "ARSA" is selected, the ordering is found by first -->
<!-- building a weighted Euclidean distance matrix, which is then passed to -->
<!-- the Simulated Annealing algorithm. Note that weighting is -->
<!-- done with respect to either the GC sites or the CG sites - the plot on -->
<!-- which brushing is performed determines which sites to use. -->


<!-- If "Refinement" is selected as the brushing option, clicking and -->
<!-- dragging on the sequence plot will select reads (i.e., rows) to -->
<!-- reorder locally. PCA is used by default, but Hierarchical Clustering -->
<!-- can also be selected as the refinement method. Unlike re-weightings, -->
<!-- refinements to the sequence plot stack onto each other, and several -->
<!-- refinements can be done to a single plot before exporting. However, it -->
<!-- is important to note that re-weighting the sites will reorder the -->
<!-- entire set of data, and hence will undo any refinements that you may -->
<!-- have made. -->

<!-- The green lines indicate the columns selected for weighting, and the -->
<!-- blue lines indicate the rows that were most recently refined. -->

<!-- After making any desired changes, the sequence plot can be saved as -->
<!-- either a PNG, PDF, or SVG file. Additionally, `methylscaper` keeps -->
<!-- track of all changes made to the plots in the form of a changes log, -->
<!-- which can be saved as a text file. -->

<!-- The "Single-molecule" panel includes a "Preprocessing" tab, where -->
<!-- users are prompted to input two sequence FASTA files, the first -->
<!-- containing all of the raw reads and second containing the reference -->
<!-- sequence for the region of interest. Users are asked to enter two -->
<!-- files paths, indicating where to save the  -->
<!-- two processed state matrices. The GCH file contains the GC site -->
<!-- accessibility data, and the HCG file contains CG site methylation data. These files can -->
<!-- then be loaded on the "Seriation" tab, which offers the same -->
<!-- functionality as described above. -->
